### PR TITLE
Add Exit system call class to Tock 2.0.

### PR DESCRIPTION
**Pull request by phil-levis**
### Pull Request Overview

This pull request adds the Exit system call class to Tock 2.0, including `exit-terminate` and `exit-restart`. It refactors some of the process restart code to separate policy from mechanism. It builds on the `yield_2.0` branch/PR #2351 . This borrows heavily from @torfmaster's #1624. 

### Testing Strategy

This pull request was tested by running the `restart` and `terminate` test apps in libtock-c. Both behave as expected, and work correctly with process console.

### TODO or Help Wanted

This pull request doesn't need anything, but process console could probably benefit from a command to reboot a process as well as one to start a terminated process.

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

File: doc/reference/trd-syscalls.md
Comment (alistair23) [inline doc/reference/trd-syscalls.md:640]: Why not at least say a completion code of 0 is success?

**Review by alistair23 — CHANGES_REQUESTED** [review]

File: doc/reference/trd-syscalls.md
Comment (phil-levis) [reply doc/reference/trd-syscalls.md:640]: Because that's an aspect of POSIX semantics which I'm not certain we want to borrow. We very well might, but we haven't fully thought this through. Rather than block this document on a design for process exit codes, I thought it better to just defer to another document. Nothing depends on or uses this yet, so binding it so early seems like a mistake.

Comment (alistair23) [reply doc/reference/trd-syscalls.md:640]: We should bind it at some point, otherwise different apps are going to return different values and it will become fragmented.

I don't think returning 0 for success is just a POSIX thing, doesn't Windows and everything else do that as well?

Just to clarify I'm not saying specify the meanings, just something along the lines of negative numbers mean failure, 0 means success (or something like that).

Comment (alistair23) [reply doc/reference/trd-syscalls.md:640]: Also, we could copy the low level debug meanings here and then remove it and replace it with this instead.

File: doc/reference/trd-syscalls.md
Comment (jrvanwhy) [inline doc/reference/trd-syscalls.md:204]: I personally found the wording "Any value not specified in the above table is reserved." a bit ambiguous. The way I read it, it was "reserved for future use" -- a future kernel version could use a different values, and processes must be ready to handle the unexpected value.

The way I read your new wording, you are saying that a Tock 2.0 kernel will *never* return a different value, which allows for a different `libtock-rs` implementation.

I do not expect that you intended this to be a semantics change, because if it is a semantics change it seems orthogonal to this PR. Should this be in a different PR?

Comment (jrvanwhy) [reply doc/reference/trd-syscalls.md:204]: Oh, this was actually part of #2351 .

File: doc/reference/trd-syscalls.md
Comment (jrvanwhy) [reply doc/reference/trd-syscalls.md:640]: I kinda think the low level debug alert codes should have been runtime-specific from the start, so I'm not sure I agree.

Comment (phil-levis) [reply doc/reference/trd-syscalls.md:640]: I agree we should bind it at some point (even very soon). I just think that this isn't the place to do it. It's going to be part of discussing, designing, and specifying the lifetime of a process in Tock.

File: doc/reference/trd-syscalls.md
Comment (jrvanwhy) [inline doc/reference/trd-syscalls.md:643]: If I'm reading the code correctly, `exit-terminate` and `exit-restart` never return regardless of whether they fail. Is that correct?

Do we want to guarantee that behavior for `exit-terminate`? It would make the entry point logic slightly simpler.

Comment (phil-levis) [reply doc/reference/trd-syscalls.md:643]: `exit-terminate` and `exit-restart` always succeed and so should never return. We should guarantee it for both of them. I think going down the "conditional restart" path is probably not helpful (e.g., "well exit restart only if you'll restart me").

Comment (phil-levis) [reply doc/reference/trd-syscalls.md]: I've updated the document to say that that `exit-restart` and `exit-terminate` MUST succeed (and so never return).

**Review by alistair23 — APPROVED** [review]

**Comment by bradjc** [comment]
This PR will be hard to reconcile with master. Can you move the restart function back to where it was (at least in git's view) so the diff is smaller?

**Comment by phil-levis** [comment]
> 
> 
> This PR will be hard to reconcile with master. Can you move the restart function back to where it was (at least in git's view) so the diff is smaller?

I don't think so, because now it's inside the `ProcessType` trait (is invoked from sched.c)? My thought is that this mechanism (the policy is in `try_restart`) should be available through the kernel. The policy in `try_restart` is in response to a userspace request. A command from `pconsole` might be handled differently.

**Comment by phil-levis** [comment]
Let's discuss this on the core call tomorrow? Since it's a major change I'd like a little discussion before merging.

**Review by alevy — APPROVED** [review]

**Comment by phil-levis** [comment]
OK, I moved the restart mechanism out of the `ProcessType` trait so that all attempts to restart a process must flow through the policy implemented in `try_restart.` I will think about a later PR that separates this out with the explanation why it is necessary to have extensible policies within the kernel.

**Review by bradjc — APPROVED** [review]

File: doc/reference/trd-syscalls.md
Comment (jrvanwhy) [reply doc/reference/trd-syscalls.md:204]: Phil, is there a merge or rebase that should be done to remove changes than have already been merged?

**Pull request merged**
